### PR TITLE
Support Graphsage

### PR DESCRIPTION
**Pull request by gpzlx1**
Not only does it support `graphsage`, but it also includes the following changes.
1. Support passes `dce` and `cse`
2. fix some bugs in `gs.jit.compile`
3. Class Matrix support `all_indices`, `columnwise_sampling` and `load_dgl_graph`
4. Graph(C++) support `_CAPI_metadata` to get tensors of csc and columns_id in python-level
5. C++ `ColumnwiseSampling` and `ColumnwiseFusedSlicingAndSampling` support `SetColIds`
6. Do fusion for columnwise_slicing and columnwise_sampling in IR level

### CODE REVIEW

**Comment by gpzlx1** [comment]
Have merged PR5;

File: python/gs/matrix_api/matrix.py [hunk 1]
```diff
@@ -6,12 +6,28 @@ def __init__(self, graph: torch.classes.gs_classes.Graph):
         # Graph bind to a C++ object
         self._graph = graph
 
-    def load_dgl_graph(self):
-        pass
+    def load_dgl_graph(self, g):
+        # import csc
+        import dgl
```
Comment (czkkkkkk) [inline python/gs/matrix_api/matrix.py:11]: Put imports at the top of a module

File: src/cuda/graph_ops.cu [hunk 2]
```diff
@@ -268,5 +281,21 @@ CSCColumnwiseFusedSlicingAndSamplingCUDA(torch::Tensor indptr,
   return {sub_indptr, sub_indices};
 }
 
+torch::Tensor TensorUniqueCUDA(torch::Tensor node_ids) {
+  return unique_single(node_ids);
+}
+
+
+std::tuple<torch::Tensor, torch::Tensor> relabelCUDA(torch::Tensor col_ids, torch::Tensor indices){
```
Comment (czkkkkkk) [inline src/cuda/graph_ops.cu:289]: RelabelCUDA

File: src/cuda/include/atomic.cuh [hunk 3]
```diff
@@ -0,0 +1,220 @@
+#pragma once
```
Comment (czkkkkkk) [inline src/cuda/include/atomic.cuh:1]: Use `#ifdef` instead of `progma once` for consistency.

File: src/cuda/static_hashmap.cu [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+  key = key ^ (key >> 14);
+  key = (key + (key << 2)) + (key << 4);    // key * 21
+  key = key ^ (key >> 28);
+  key = key + (key << 31);
+  return key;
+}
+
+
+template<typename K, typename V, typename Op>
+struct static_hashmap
```
Comment (czkkkkkk) [inline src/cuda/static_hashmap.cu:39]: StaticHashmap

File: src/cuda/static_hashmap.cu [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+
+
+int64_t UpPower(int64_t key){
+    int64_t ret =  1 << static_cast<uint64_t>(std::log2(key) + 1);
+    return ret;
+}
+
+/******************************* shifthash **************************************/
+/**************** Thomas Wang's 32 bit and 64 bit Mix Function ******************/
+__device__ inline uint32_t hash32shift(uint32_t key){
```
Comment (czkkkkkk) [inline src/cuda/static_hashmap.cu:14]: Hash32Shift

File: src/cuda/static_hashmap.cu [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+
+    K kEmptyKey;
+    V DefaultValue;
+    K* kptr;
+    V* vptr;
+    uint32_t capacity{0};
+};
+
+template<typename KType, typename VType, typename Op>
+void static_hashmap_update_core(
```
Comment (czkkkkkk) [inline src/cuda/static_hashmap.cu:123]: StaticHashmapUpdateCore

File: src/cuda/unique_reindex.cu [hunk 7]
```diff
@@ -0,0 +1,347 @@
+#include "cuda_ops.cuh"
+#include "atomic.cuh"
+
+
+template <typename IdType>
+struct relabel_hashmap
```
Comment (czkkkkkk) [inline src/cuda/unique_reindex.cu:6]: RelabelHashmap

File: src/graph.h [hunk 8]
```diff
@@ -14,15 +14,21 @@ class Graph : public torch::CustomClassHolder {
...
   torch::Tensor RowIndices();
+  torch::Tensor AllIndices();
+  std::tuple<torch::Tensor, torch::Tensor, torch::Tensor> Relabel();
+  std::vector<torch::Tensor> MetaData();
 
   void Print() const;
 
  private:
   bool is_subgraph_;
+  torch::Tensor _col_ids; //used for subgraoh, stores node id in global graph
```
Comment (czkkkkkk) [inline src/graph.h:31]: Can we put the `col_ids` in the CSC?

File: src/graph.cc [hunk 9]
```diff
@@ -12,20 +10,30 @@ void Graph::LoadCSC(torch::Tensor indptr, torch::Tensor indices) {
   csc_->indices = indices;
 }
 
+std::shared_ptr<CSC> Graph::GetCSC() { return csc_; }
+
 void Graph::SetCSC(std::shared_ptr<CSC> csc) { csc_ = csc; }
 
-std::shared_ptr<CSC> Graph::GetCSC() { return csc_; }
+void Graph::SetColIds(torch::Tensor column_ids){
```
Comment (czkkkkkk) [inline src/graph.cc:17]: col_ids_

File: src/graph.h [hunk 10]
```diff
@@ -14,15 +14,21 @@ class Graph : public torch::CustomClassHolder {
...
   void SetCSC(std::shared_ptr<CSC> csc);
   std::shared_ptr<CSC> GetCSC();
+  torch::Tensor GetcolIds();
+  void SetColIds(torch::Tensor column_ids);
   c10::intrusive_ptr<Graph> ColumnwiseSlicing(torch::Tensor column_ids);
   c10::intrusive_ptr<Graph> ColumnwiseSampling(int64_t fanout, bool replace);
   c10::intrusive_ptr<Graph> ColumnwiseFusedSlicingAndSampling(torch::Tensor column_ids, int64_t fanout, bool replace);
   torch::Tensor RowIndices();
+  torch::Tensor AllIndices();
+  std::tuple<torch::Tensor, torch::Tensor, torch::Tensor> Relabel();
```
Comment (czkkkkkk) [inline src/graph.h:24]: Could you give some explanations about what does this interface do? What are the three returned tensors?

File: src/cuda/unique_reindex.cu [hunk 11]
```diff
@@ -0,0 +1,347 @@
...
+){
+    torch::Tensor total_tensor = data;
+    std::vector<torch::Tensor> ret;
+    OGS_ID_TYPE_SWITCH(total_tensor.dtype(), IdType, {
+        ret = unique_core<IdType, true>(total_tensor);
+    });
+    return ret;
+}
+
+std::tuple<torch::Tensor, std::vector<torch::Tensor>> relabel(
```
Comment (czkkkkkk) [inline src/cuda/unique_reindex.cu:264]: Relabel

**Review by czkkkkkk — COMMENTED** [review]
Overall LGTM, but I still have several concerns.

1. Why we need to have an include directory in the `src/cuda`?
2. Give some explanations about the `Graph::relabel()` interface.
3. Naming and formatting convention. Python code is fine but we need to have the same coding style for C++/CUDA code. I suggest to follow Google coding style (https://google.github.io/styleguide/cppguide.html) especially for the naming rules. For code formatting, we have `.clang-format` in the src directory, which will be used automatically if IDEs (e.g., vscode) try to format the code. We can also add a linting checker later but now only using IDE auto formatters is fine.

**Comment by czkkkkkk** [comment]
BTW, can we wrap a Pytorch Dataloader to sample the nodes in an epoch?

**Comment by gpzlx1** [comment]
It's possible. However, my idea is to implement a `seedGenerater`. 
Users call our sampling function in the following ways for better performance.
```python
for seeds in seedGenerater(train_node, batch_size, shuffle=True):
    graph = compiled_func(m, seeds, [25, 10])
```

**Comment by gpzlx1** [comment]
@czkkkkkk Removed unnecessary code, renamed variables, formatted some code, added comments to Graph::Relabel, Graph::All_Indices.

File: src/graph.cc [hunk 12]
```diff
@@ -39,8 +44,50 @@ c10::intrusive_ptr<Graph> Graph::ColumnwiseFusedSlicingAndSampling(
...
+ * csc_indptr of the relabeled graph; new_csc_indices is the csc_indices of the
+ * relabeled graph.
+ * For example,
+ *    if graph.csc_.col_ids = [0, 2, 4, 2], graph.csc_.indptr = [0, 0, 1, 1, 3]
+ *    and graph.csc_.indices = [4, 2, 1], 
+ *    graph.relabel will return {[0, 2, 4, 1], [0, 0, 1, 1, 3], [2, 1, 3]}
+ *
+ * @return std::tuple<torch::Tensor, torch::Tensor, torch::Tensor>
+ */
+std::tuple<torch::Tensor, torch::Tensor, torch::Tensor> Graph::Relabel(){
```
Comment (czkkkkkk) [inline src/graph.cc:82]: I wonder whether we need to store row_ids after relabel. The row_ids may be used to fetch features.

File: src/graph.h [hunk 13]
```diff
@@ -12,17 +12,21 @@ class Graph : public torch::CustomClassHolder {
...
+  std::vector<torch::Tensor> MetaData();
 
   void Print() const;
 
  private:
-  bool is_subgraph_;
+  bool is_subgraph_; //used for subgraoh, stores node id in global graph
```
Comment (czkkkkkk) [inline src/graph.h:29]: typo

**Review by czkkkkkk — APPROVED** [review]

File: src/graph.cc [hunk 14]
```diff
@@ -39,8 +44,50 @@ c10::intrusive_ptr<Graph> Graph::ColumnwiseFusedSlicingAndSampling(
...
+ * @brief Returns the set of all nodes of the graph. Nodes in return tensor are
+ * sorted in the order of their first occurrence in {col_ids, indices}.
+ * For example, 
+ *    if graph.csc_.col_ids = [0, 2, 4, 2] and graph.csc_.indices = [4, 2, 1], 
+ *    graph.AllIndices() will be [0, 2, 4, 1]
+ *
+ * @return torch::Tensor
+ */
+torch::Tensor Graph::AllIndices() {
+  if(is_subgraph_){
```
[the remark this answers is not in the document]
Comment (gpzlx1) [reply src/graph.cc:57]: When this PR merge, I plan to open another PR for formatting all files and public header file extraction.

File: src/graph.cc (hunk 12 above)
Comment (gpzlx1) [reply src/graph.cc:82]: We only call Graph::Relabel when we need Graph to return. The result of Graph::Relabel contains the information of `allindices`, which is used to fetch the feature. The relabel operation for heterogeneous graphs may be very different and this API may not be applicable. Let's discuss later.

**Pull request merged**